### ISSUE TITLE
Incorrect reference counting of 'yara.StringMatchInstance' attributes

### ISSUE THREAD

**Issue by Dadeos-Menlo** [VirusTotal/yara-python#242]
There appears to be a few issues with Python Object reference counting associated with the [yara.StringMatchInstance](https://yara.readthedocs.io/en/v4.3.0/yarapython.html#yara.StringMatchInstance) objects introduced in version 4.3.0.

The main issue regards the reference counting, or lack thereof, of the return value provided by the [yara.StringMatchInstance.plaintext()](https://yara.readthedocs.io/en/v4.3.0/yarapython.html#yara.StringMatchInstance.plaintext) function in scenarios where an [xor_key](https://yara.readthedocs.io/en/v4.3.0/yarapython.html#yara.StringMatchInstance.xor_key) is not in use (i.e. `xor_key == 0`):

https://github.com/VirusTotal/yara-python/blob/a4b2ae23b92fbd555e35d7bed3dcf218526b2388/yara-python.c#L1918-L1929

It can be seen that under such circumstances the return value from the function is the Python Object referred to by `instance->matched_data`. However, the reference count on this Python Object has not been [incremented](https://docs.python.org/3/c-api/refcounting.html#c.Py_INCREF), and is consequently freed multiple times generally leading to heap corruption.

This issue may be observed using the following unit test case:
```python
import unittest

import yara

class TestYARA(unittest.TestCase):

   def test_ref_count_plaintext(self):
      count = 100
      iterations = 100
      rules = yara.compile(
         source='rule test { strings: $string = {01} condition: $string }'
      )
      data = b'\x00\x01' * count

      for iteration in range(iterations):
         for match in rules.match(data=data):
            for string in match.strings:
               for instance in string.instances:
                  instance.plaintext()

if __name__ == "__main__":
    unittest.main()
```
```
$ bash -c 'set -e; python3 test.py')
bash: line 1: 10118 Segmentation fault      (core dumped) python3 test.py
```
Incrementing the reference count of the `instance->matched_data` object prior to returning a reference to it resolves the issue:
```c
  if (xor_key == 0)
  {
      Py_INCREF(instance->matched_data);
      return instance->matched_data;
  }
```

During the investigation of the heap corruption I also observed a memory leak, caused by incorrect reference counting, associated with the [yara.StringMatchInstance.matched_length](https://yara.readthedocs.io/en/v4.3.0/yarapython.html#yara.StringMatchInstance.matched_length) attribute. The `matched_length` attribute is a Python Object created using [PyLong_FromLong](https://docs.python.org/3/c-api/long.html#c.PyLong_FromLong):

https://github.com/VirusTotal/yara-python/blob/a4b2ae23b92fbd555e35d7bed3dcf218526b2388/yara-python.c#L1872

However, its reference count is not decremented in the deallocation function:

https://github.com/VirusTotal/yara-python/blob/a4b2ae23b92fbd555e35d7bed3dcf218526b2388/yara-python.c#L1882-L1892

alongside the other attributes, whose reference counts are decremented.

The resultant memory leak may be observed using the following unit test case:
```python
import inspect
import tracemalloc
import unittest

import yara

class TestYARA(unittest.TestCase):

   def test_ref_count_matched_length_leak(self):
      frame = inspect.getframeinfo(inspect.currentframe())

      count = 100
      length = 257
      rules = yara.compile(
         source='rule test { strings: $string = {' +
            ('01 ' * length) +
         '} condition: $string }'
      )
      data = (b'\x00' + b'\x01' * length) * count

      tracemalloc.start()
      before = tracemalloc.take_snapshot()
      rules.match(data=data)
      after = tracemalloc.take_snapshot()
      tracemalloc.stop()

      self.assertEqual([], [entry
         for entry in after.compare_to(before, 'lineno')
            if entry.traceback[0].filename == __file__
      ])

if __name__ == "__main__":
    unittest.main()
``` 
```
#  python3 test.py
F
======================================================================
FAIL: test_ref_count_matched_length_leak (__main__.TestYARA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test.py", line 27, in test_ref_count_matched_length_leak
    self.assertEqual([], [entry
AssertionError: Lists differ: [] != [<StatisticDiff traceback=<Traceback (<Fra[77 chars]00)>]

Second list contains 1 additional elements.
First extra element 0:
<StatisticDiff traceback=<Traceback (<Frame filename='test.py' lineno=23>,)> size=2800 (+2800) count=100 (+100)>

- []
+ [<StatisticDiff traceback=<Traceback (<Frame filename='test.py' lineno=23>,)> size=2800 (+2800) count=100 (+100)>]

----------------------------------------------------------------------
Ran 1 test in 0.013s

FAILED (failures=1)
```
Decrementing the reference count of the `instance->matched_length` object in the deallocation function resolves the issue:
```c
static void StringMatchInstance_dealloc(
    PyObject* self)
{
  StringMatchInstance* object = (StringMatchInstance*) self;

  Py_DECREF(object->offset);
  Py_DECREF(object->matched_data);
  Py_DECREF(object->matched_length);
  Py_DECREF(object->xor_key);

  PyObject_Del(self);
}
```

As an aside; the [yara.StringMatchInstance.matched_length](https://yara.readthedocs.io/en/v4.3.0/yarapython.html#yara.StringMatchInstance.matched_length) attribute appears to be somewhat redundant, given that I believe that its value will always match the length of the [yara.StringMatchInstance.matched_data](https://yara.readthedocs.io/en/v4.3.0/yarapython.html#yara.StringMatchInstance.matched_data) attribute:
```python
   instance.matched_length == len(instance.matched_data)
```

**Comment by plusvic** [comment]
Already fixed in #243.

**Comment by Dadeos-Menlo** [comment]
Are there any plans to release a 4.3.2 version containing this fix?

**Comment by plusvic** [comment]
This will be released in the upcoming 4.4.0 release.
